docs/devel/controllers.md

```diff
@@ -0,0 +1,101 @@
+
+
+1. Use `SharedInformers`.  `SharedInformers` provide hooks to receive notifications of adds, updates, and deletes for a particular resource.  They also provide convenience functions for accessing shared caches and determining when a cache is primed.
+
+ Use the factory methods down in https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/framework/informers/factory.go to ensure that you are sharing the same instance of the cache as everyone else.
+
+ This saves us connections against the API server, duplicate serialization costs server-side, duplicate deserialization costs controller-side, and duplicate caching costs controller-side.
+
+ You may see other mechanisms like reflectors and deltafifos driving controllers.  Those were older mechanisms that we later used to build the `SharedInformers`.  You should avoid using them in new controllers
+
+
+1. Never mutate original objects!  Caches are shared across controllers, this means that if you mutate your copy of an object, you’ll mess up other controllers (not just your own).  
+
+ The most common point of failure is making a shallow copy, then mutating a map, like `Annotations`.  Use `api.Scheme.Copy` to make a deep copy.
+
+
+1. Wait for your secondary caches.  Many controllers have primary and secondary resources.  Primary resources are the resources that you’ll be updating `Status` for.  Secondary resources are resources that you’ll be managing (creating/deleting) or using for lookups.
+
+ Use the `framework.WaitForCacheSync` function to wait for your secondary caches before starting your primary sync functions.  This will make sure that things like a Pod count for a ReplicaSet isn’t working off of known out of date information that results in thrashing.
+
+
+1. Percolate errors to the top level for consistent re-queuing.  We have a  `workqueue.RateLimitingInterface` to allow simple requeuing with reasonable backoffs.
+
+ Your main controller func should return an error when requeuing is necessary.  When it isn’t, it should se `utilruntime.HandleError` and return nil instead.  This makes it very easy for reviewers to inspect error handling cases and to be confident that your controller doesn’t accidentally lose things it should retry for.
+
+
+1. Watches and Informers will “sync”.  Periodically, they will deliver every matching object in the cluster to your `Update` method.  This is good for cases where you may need to take additional action on the object, but sometimes you know there won’t be more work to do.
+
+ In cases where you are *certain* that you always requeue failing items, you can compare the resource version of the old and new objects.  If they are the same, you skip requeuing the work.
```

Comment:
"failing items" doesn't make any sense to me.  How about "In case where you are certain that you don't need to requeue items with no new changes" or something similar